pkg/autoscaler/statforwarder/forwarder.go

```diff
@@ -0,0 +1,229 @@
+			// TODO(yanweiguo): Set up DeleteFunc.
+		},
+	})
+
+	return &f
+}
+
+func (f *Forwarder) leaseUpdated(obj interface{}) {
+	l := obj.(*coordinationv1.Lease)
+	ns, n := l.Namespace, l.Name
+
+	if !f.bs.HasBucket(n) {
+		// Skip this Lease as it is not for Autoscaler.
+		return
+	}
+
+	if l.Spec.HolderIdentity == nil || *l.Spec.HolderIdentity == "" {
+		f.logger.Debugf("HolderIdentity not found for Lease %s/%s", ns, n)
+		return
+	}
+
+	identity := *l.Spec.HolderIdentity
+	if h, ok := f.getHolder(n); ok && h == identity {
+		// Already up-to-date.
+		return
+	}
+
+	f.setHolder(n, identity)
+
+	if identity != f.selfIP {
+		// Skip creating/updating Service and Endpoints as not the holder.
+		return
+	}
+
+	if err := f.createService(ns, n); err != nil {
+		f.logger.Errorf("Failed to create Service for Lease %s/%s: %v", ns, n, err)
+		return
+	}
+	f.logger.Infof("Created Service for Lease %s/%s", ns, n)
+
+	if err := f.createOrUpdateEndpoints(ns, n); err != nil {
+		f.logger.Errorf("Failed to create Endpoints for Lease %s/%s: %v", ns, n, err)
+		return
+	}
+	f.logger.Infof("Created Endpoints for Lease %s/%s", ns, n)
+}
+
+func (f *Forwarder) createService(ns, n string) error {
+	return wait.ExponentialBackoff(retry.DefaultRetry, func() (bool, error) {
+		_, err := f.kc.CoreV1().Services(ns).Create(&v1.Service{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      n,
+				Namespace: ns,
+			},
+			Spec: v1.ServiceSpec{
+				Ports: []v1.ServicePort{{
+					Name:       autoscalerPortName,
+					Port:       autoscalerPort,
+					TargetPort: intstr.FromInt(autoscalerPort),
+				}},
+			},
+		})
+
+		if apierrs.IsAlreadyExists(err) {
+			return true, nil
+		}
+
+		return err == nil, err
+	})
+}
+
+// createOrUpdateEndpoints creates a Endpoints object with the given namespace and
+// name, and the Forwarder.selfIP. If the Endpoints object already
+// exists, it will update the Endpoints with the Forwarder.selfIP.
+func (f *Forwarder) createOrUpdateEndpoints(ns, n string) error {
+	wantSubsets := []v1.EndpointSubset{{
+		Addresses: []v1.EndpointAddress{{
+			IP: f.selfIP,
+		}},
+		Ports: []v1.EndpointPort{{
+			Name:     autoscalerPortName,
+			Port:     autoscalerPort,
+			Protocol: v1.ProtocolTCP,
+		}}},
+	}
+
+	exists := true
+	if err := wait.ExponentialBackoff(retry.DefaultRetry, func() (bool, error) {
+		e, err := f.endpointsLister.Endpoints(ns).Get(n)
+
+		if apierrs.IsNotFound(err) {
+			exists = false
+			return true, nil
+		}
+
+		if err != nil {
+			return false, err
+		}
+
+		if equality.Semantic.DeepEqual(wantSubsets, e.Subsets) {
+			return true, nil
+		}
+
+		want := e.DeepCopy()
+		want.Subsets = wantSubsets
+		if _, err := f.kc.CoreV1().Endpoints(ns).Update(want); err != nil {
+			return false, err
+		}
+
+		f.logger.Infof("Bucket Endpoints %s updated with IP %s", n, f.selfIP)
+		return true, nil
+	}); err != nil {
+		return err
+	}
+
+	if exists {
+		return nil
+	}
+	return wait.ExponentialBackoff(retry.DefaultRetry, func() (bool, error) {
```

Comment:
I thought we already have this wrapped in pkg somewhere? pkg/reconciler?

Reply:
Didn't find a one.

Reply:
https://github.com/knative/pkg/blob/master/reconciler/retry.go#L24
I meant this one

Reply:
That one checks on Conflict error. Could create return other errors we should retry?

Reply:
AlreadyExists is also the one that can come out here.

Reply:
Do you mean only Conflict or AlreadyExists can happen here? If so, we should retry on Conflict but not on AlreadyExists. https://github.com/knative/pkg/blob/master/reconciler/retry.go#L24 can be used.

Could there be other errors, for example, timeout due to temporary network issue? If so, we should retry and https://github.com/knative/pkg/blob/master/reconciler/retry.go#L24 can't be used.

Reply:
I concur. It's fine the way it is .